R/asmodee.R

```diff
@@ -8,332 +8,411 @@
+        "Consider using `include_prediction_warnings = TRUE`",
+        "to include more models which issued warnings.",
+        sep = "\n"
+      )
+    }
+    stop(msg)
   }
 
+  # subset to the best model and then combine with prediction results
+  out <- out[i, ]
+  models <- models[i]
+  out <- dplyr::bind_cols(out, pred_result)
 
-  ## find outliers
-  res_outliers <- detect_outliers(data = data,
-                                  model = selected_model,
-                                  alpha = alpha,
-                                  simulate_pi = simulate_pi,
-                                  uncertain = uncertain)
+  # mark up outliers (note the hacks needed for precision issues)
+  preds <- out$result[[1]]
+  model <- out$trending_model_fit[[1]]$fitted_model
+  observed <- all.vars(formula(model))[1]
+  preds <- dplyr::mutate(
+    preds,
+    outlier = .data[[observed]] < .data$lower_pi | .data[[observed]] > .data$upper_pi,
+    outlier = dplyr::case_when(
+      outlier & dplyr::near(.data[[observed]], .data$lower_pi) ~ FALSE,
+      outlier & dplyr::near(.data[[observed]], .data$upper_pi) ~ FALSE,
+      TRUE ~ outlier
+    ),
+    classification = dplyr::case_when(
+      (.data[[observed]] < .data$lower_pi) & .data$outlier ~ "decrease",
+      (.data[[observed]] > .data$upper_pi) & .data$outlier ~ "increase",
+      TRUE ~ "normal"
+    ),
+    classification = factor(
+      .data$classification,
+      levels = c("increase", "normal", "decrease")
+    )
+  )
 
-  ## enforce positive predictions if required
-  neg_to_zero <- function(x) {
-    x[x < 0] <- 0
-    x
-  }    
+  # enforce positive predictions if required
   if (force_positive) {
-    res_outliers$estimate <- neg_to_zero(res_outliers$estimate)
-    res_outliers$lower_ci <- neg_to_zero(res_outliers$lower_ci)
-    res_outliers$upper_ci <- neg_to_zero(res_outliers$upper_ci)
-    res_outliers$lower_pi <- neg_to_zero(res_outliers$lower_pi)
-    res_outliers$upper_pi <- neg_to_zero(res_outliers$upper_pi)
+    nms <- c("estimate", "lower_ci", "upper_ci", "lower_pi", "upper_pi")
+    neg_to_zero <- function(x) {
+      x[x < 0] <- 0
+      x
+    }
+    preds[nms] <- lapply(preds[nms], neg_to_zero)
   }
 
-  ## final output
+  # final output
   out <- list(
-    k = selected_k,
-    model = selected_model,
+    k = k,
+    model_name = out$model_name[[1]],
+    trending_model_fit = out$trending_model_fit[[1]],
     alpha = alpha,
-    results = res_outliers,
+    results = tibble::as_tibble(preds),
     date_index = date_index,
     last_training_date = last_training_date,
-    first_testing_date = first_testing_date
+    first_testing_date = first_testing_date,
+    .fitted_results = .fitted_results
   )
   class(out) <- c("trendbreaker", class(out))
   out
 }
 
 
-
-
-
-#' @export
-#'
 #' @rdname asmodee
+#' @param include_group_warnings Should results include models that triggered
+#'   warnings, but not errors, during the prediciton stage. Defaults to `FALSE`.
 #'
```

Comment:
I have changed this description slightly, but need to check new version is correct; it looks like this will catch warning/errors from `asmodee.data.frame`, but I am not sure what warnings would be, it does not look like we are generating any (or is it capturing all warnings from called sub-procedures)? Also, we might want the default to be FALSE? Merely putting this as a discussion item.

Reply:
Good point - let's remove this check for the time being.